File: irsp/codec/msgpack_codec.cpp
#include "codec/msgpack_codec.hpp"

#include <algorithm>
#include <cstdint>
#include <cstring>
#include <variant>

#include "codec/irsp1_codec.hpp"

namespace irsp {

namespace {

// ============================ 编码 ============================

void putU8(std::string &o, std::uint8_t b) { o.push_back(static_cast<char>(b)); }

/// 追加 v 的低 n 字节，大端序（msgpack 多字节整数恒为大端）。
void putBE(std::string &o, std::uint64_t v, int n) {
    for (int i = n - 1; i >= 0; --i) {
        o.push_back(static_cast<char>((v >> (8 * i)) & 0xff));
    }
}

void encUint(std::string &o, std::uint64_t u) {
    if (u <= 0x7f) {
        putU8(o, static_cast<std::uint8_t>(u)); // positive fixint
    } else if (u <= 0xff) {
        putU8(o, 0xcc);
        putBE(o, u, 1);
    } else if (u <= 0xffff) {
        putU8(o, 0xcd);
        putBE(o, u, 2);
    } else if (u <= 0xffffffffULL) {
        putU8(o, 0xce);
        putBE(o, u, 4);
    } else {
        putU8(o, 0xcf);
        putBE(o, u, 8);
    }
}

void encSint(std::string &o, std::int64_t v) {
    if (v >= 0) {
        encUint(o, static_cast<std::uint64_t>(v));
    } else if (v >= -32) {
        putU8(o, static_cast<std::uint8_t>(v)); // negative fixint（两补码低 8 位即 0xe0..0xff）
    } else if (v >= -128) {
        putU8(o, 0xd0);
        putBE(o, static_cast<std::uint64_t>(v), 1);
    } else if (v >= -32768) {
        putU8(o, 0xd1);
        putBE(o, static_cast<std::uint64_t>(v), 2);
    } else if (v >= -2147483648LL) {
        putU8(o, 0xd2);
        putBE(o, static_cast<std::uint64_t>(v), 4);
    } else {
        putU8(o, 0xd3);
        putBE(o, static_cast<std::uint64_t>(v), 8);
    }
}

void encStr(std::string &o, const std::string &s) {
    const std::size_t n = s.size();
    if (n <= 31) {
        putU8(o, static_cast<std::uint8_t>(0xa0 | n)); // fixstr
    } else if (n <= 0xff) {
        putU8(o, 0xd9);
        putBE(o, n, 1);
    } else if (n <= 0xffff) {
        putU8(o, 0xda);
        putBE(o, n, 2);
    } else {
        putU8(o, 0xdb);
        putBE(o, n, 4);
    }
    o.append(s);
}

void encBin(std::string &o, const std::string &s) {
    const std::size_t n = s.size();
    if (n <= 0xff) {
        putU8(o, 0xc4);
        putBE(o, n, 1);
    } else if (n <= 0xffff) {
        putU8(o, 0xc5);
        putBE(o, n, 2);
    } else {
        putU8(o, 0xc6);
        putBE(o, n, 4);
    }
    o.append(s);
}

void encArrHdr(std::string &o, std::size_t n) {
    if (n <= 15) {
        putU8(o, static_cast<std::uint8_t>(0x90 | n)); // fixarray
    } else if (n <= 0xffff) {
        putU8(o, 0xdc);
        putBE(o, n, 2);
    } else {
        putU8(o, 0xdd);
        putBE(o, n, 4);
    }
}

void encMapHdr(std::string &o, std::size_t n) {
    if (n <= 15) {
        putU8(o, static_cast<std::uint8_t>(0x80 | n)); // fixmap
    } else if (n <= 0xffff) {
        putU8(o, 0xde);
        putBE(o, n, 2);
    } else {
        putU8(o, 0xdf);
        putBE(o, n, 4);
    }
}

/// 从小端原始字节读 n 字节为无符号整数。
std::uint64_t readLE(const std::string &r, std::size_t n) {
    std::uint64_t v = 0;
    for (std::size_t i = 0; i < n; ++i) {
        v |= static_cast<std::uint64_t>(static_cast<std::uint8_t>(r[i])) << (8 * i);
    }
    return v;
}

/// 依类型标签把 TagValue.value（小端原始字节）编码为 msgpack 原生类型。
/// 类型未知或字节长度不符 -> 原样 bin 兜底（无损）。
void encTyped(std::string &o, const IrspTypedValue &tv) {
    const std::string &t = tv.type;
    const std::string &r = tv.raw;
    const std::size_t sz = r.size();

    if (t == "null") {
        putU8(o, 0xc0);
    } else if (t == "bool") {
        putU8(o, (sz >= 1 && r[0] != 0) ? 0xc3 : 0xc2);
    } else if (t == "str") {
        encStr(o, r);
    } else if (t == "i8" && sz >= 1) {
        encSint(o, static_cast<std::int8_t>(readLE(r, 1)));
    } else if (t == "i16" && sz >= 2) {
        encSint(o, static_cast<std::int16_t>(readLE(r, 2)));
    } else if (t == "i32" && sz >= 4) {
        encSint(o, static_cast<std::int32_t>(readLE(r, 4)));
    } else if (t == "i64" && sz >= 8) {
        encSint(o, static_cast<std::int64_t>(readLE(r, 8)));
    } else if (t == "u8" && sz >= 1) {
        encUint(o, readLE(r, 1));
    } else if (t == "u16" && sz >= 2) {
        encUint(o, readLE(r, 2));
    } else if (t == "u32" && sz >= 4) {
        encUint(o, readLE(r, 4));
    } else if (t == "u64" && sz >= 8) {
        encUint(o, readLE(r, 8));
    } else if (t == "f32" && sz >= 4) {
        const auto u = static_cast<std::uint32_t>(readLE(r, 4));
        float f = 0;
        std::memcpy(&f, &u, 4);
        std::uint32_t bits = 0;
        std::memcpy(&bits, &f, 4);
        putU8(o, 0xca);
        putBE(o, bits, 4);
    } else if (t == "f64" && sz >= 8) {
        const std::uint64_t u = readLE(r, 8);
        double d = 0;
        std::memcpy(&d, &u, 8);
        std::uint64_t bits = 0;
        std::memcpy(&bits, &d, 8);
        putU8(o, 0xcb);
        putBE(o, bits, 8);
    } else {
        encBin(o, r);
    }
}

void encodeInto(const IrspValue &v, std::string &o) {
    std::visit(
        [&o](const auto &x) {
            using T = std::decay_t<decltype(x)>;
            if constexpr (std::is_same_v<T, IrspNull>) {
                putU8(o, 0xc0);
            } else if constexpr (std::is_same_v<T, IrspSimple>) {
                encStr(o, x.text);
            } else if constexpr (std::is_same_v<T, IrspError>) {
                encMapHdr(o, 2); // {err: CODE, msg: MESSAGE}
                encStr(o, "err");
                encStr(o, x.code);
                encStr(o, "msg");
                encStr(o, x.message);
            } else if constexpr (std::is_same_v<T, IrspInteger>) {
                encSint(o, x.value);
            } else if constexpr (std::is_same_v<T, IrspBulk>) {
                // 编码方向 bulk 仅承载文本（topic 名 / 游标 / PING 回显 / map key）。
                encStr(o, x.data);
            } else if constexpr (std::is_same_v<T, IrspTypedValue>) {
                encTyped(o, x);
            } else if constexpr (std::is_same_v<T, IrspArray>) {
                encArrHdr(o, x.items.size());
                for (const auto &it : x.items) {
                    encodeInto(it, o);
                }
            } else if constexpr (std::is_same_v<T, IrspMap>) {
                encMapHdr(o, x.entries.size());
                for (const auto &[k, val] : x.entries) {
                    encodeInto(k, o);
                    encodeInto(val, o);
                }
            }
        },
        v);
}

// ============================ 解码 ============================

struct PResult {
    CodecStatus status{CodecStatus::Error};
    IrspValue value{IrspNull{}};
    std::size_t pos{0};
};

/// 从大端字节读 n 字节为无符号整数。
std::uint64_t rdBE(std::string_view b, std::size_t pos, int n) {
    std::uint64_t v = 0;
    for (int i = 0; i < n; ++i) {
        v = (v << 8) | static_cast<std::uint8_t>(b[pos + i]);
    }
    return v;
}

PResult parseOne(std::string_view b, std::size_t pos) {
    if (pos >= b.size()) {
        return {CodecStatus::Incomplete, IrspNull{}, pos};
    }
    const auto c = static_cast<std::uint8_t>(b[pos]);

    // str/bin：数据从 dataStart 起 len 字节，统一解码为 IrspBulk。
    auto bytes = [&](std::size_t dataStart, std::uint64_t len) -> PResult {
        if (b.size() < dataStart + len) {
            return {CodecStatus::Incomplete, IrspNull{}, pos};
        }
        return {CodecStatus::Ok, makeBulk(std::string(b.substr(dataStart, len))), dataStart + len};
    };
    // float：大端 bits -> 小端原始字节（与 IRSP value 字节约定一致），存入 IrspBulk。
    auto floatBulk = [&](std::size_t dataStart, int n) -> PResult {
        if (b.size() < dataStart + static_cast<std::size_t>(n)) {
            return {CodecStatus::Incomplete, IrspNull{}, pos};
        }
        const std::uint64_t bits = rdBE(b, dataStart, n);
        std::string le(static_cast<std::size_t>(n), '\0');
        for (int i = 0; i < n; ++i) {
            le[i] = static_cast<char>((bits >> (8 * i)) & 0xff);
        }
        return {CodecStatus::Ok, makeBulk(std::move(le)), dataStart + static_cast<std::size_t>(n)};
    };
    auto array = [&](std::size_t hdrEnd, std::uint64_t count) -> PResult {
        IrspArray arr;
        arr.items.reserve(static_cast<std::size_t>(std::min<std::uint64_t>(count, b.size())));
        std::size_t cur = hdrEnd;
        for (std::uint64_t i = 0; i < count; ++i) {
            PResult e = parseOne(b, cur);
            if (e.status != CodecStatus::Ok) {
                return {e.status, IrspNull{}, pos};
            }
            arr.items.push_back(std::move(e.value));
            cur = e.pos;
        }
        return {CodecStatus::Ok, std::move(arr), cur};
    };
    auto map = [&](std::size_t hdrEnd, std::uint64_t count) -> PResult {
        IrspMap m;
        m.entries.reserve(static_cast<std::size_t>(std::min<std::uint64_t>(count, b.size())));
        std::size_t cur = hdrEnd;
        for (std::uint64_t i = 0; i < count; ++i) {
            PResult k = parseOne(b, cur);
            if (k.status != CodecStatus::Ok) {
                return {k.status, IrspNull{}, pos};
            }
            cur = k.pos;
            PResult v = parseOne(b, cur);
            if (v.status != CodecStatus::Ok) {
                return {v.status, IrspNull{}, pos};
            }
            cur = v.pos;
            m.entries.emplace_back(std::move(k.value), std::move(v.value));
        }
        return {CodecStatus::Ok, std::move(m), cur};
    };
    // 读取 n 字节长度前缀（位于 pos+1），不足则 false。
    auto lenPrefix = [&](int n, std::uint64_t &len) -> bool {
        if (b.size() < pos + 1 + static_cast<std::size_t>(n)) {
            return false;
        }
        len = rdBE(b, pos + 1, n);
        return true;
    };

    if (c <= 0x7f) {
        return {CodecStatus::Ok, makeInteger(c), pos + 1}; // positive fixint
    }
    if (c >= 0xe0) {
        return {CodecStatus::Ok, makeInteger(static_cast<std::int8_t>(c)), pos + 1}; // neg fixint
    }
    if (c >= 0x80 && c <= 0x8f) {
        return map(pos + 1, c & 0x0f); // fixmap
    }
    if (c >= 0x90 && c <= 0x9f) {
        return array(pos + 1, c & 0x0f); // fixarray
    }
    if (c >= 0xa0 && c <= 0xbf) {
        return bytes(pos + 1, c & 0x1f); // fixstr
    }

    std::uint64_t len = 0;
    switch (c) {
    case 0xc0:
        return {CodecStatus::Ok, makeNull(), pos + 1};
    case 0xc2:
        return {CodecStatus::Ok, makeInteger(0), pos + 1};
    case 0xc3:
        return {CodecStatus::Ok, makeInteger(1), pos + 1};
    case 0xcc:
        if (b.size() < pos + 2) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int64_t>(rdBE(b, pos + 1, 1))), pos + 2};
    case 0xcd:
        if (b.size() < pos + 3) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int64_t>(rdBE(b, pos + 1, 2))), pos + 3};
    case 0xce:
        if (b.size() < pos + 5) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int64_t>(rdBE(b, pos + 1, 4))), pos + 5};
    case 0xcf:
        if (b.size() < pos + 9) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int64_t>(rdBE(b, pos + 1, 8))), pos + 9};
    case 0xd0:
        if (b.size() < pos + 2) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int8_t>(rdBE(b, pos + 1, 1))), pos + 2};
    case 0xd1:
        if (b.size() < pos + 3) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int16_t>(rdBE(b, pos + 1, 2))), pos + 3};
    case 0xd2:
        if (b.size() < pos + 5) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int32_t>(rdBE(b, pos + 1, 4))), pos + 5};
    case 0xd3:
        if (b.size() < pos + 9) break;
        return {CodecStatus::Ok, makeInteger(static_cast<std::int64_t>(rdBE(b, pos + 1, 8))), pos + 9};
    case 0xca:
        return floatBulk(pos + 1, 4);
    case 0xcb:
        return floatBulk(pos + 1, 8);
    case 0xc4:
        if (!lenPrefix(1, len)) break;
        return bytes(pos + 2, len);
    case 0xc5:
        if (!lenPrefix(2, len)) break;
        return bytes(pos + 3, len);
    case 0xc6:
        if (!lenPrefix(4, len)) break;
        return bytes(pos + 5, len);
    case 0xd9:
        if (!lenPrefix(1, len)) break;
        return bytes(pos + 2, len);
    case 0xda:
        if (!lenPrefix(2, len)) break;
        return bytes(pos + 3, len);
    case 0xdb:
        if (!lenPrefix(4, len)) break;
        return bytes(pos + 5, len);
    case 0xdc:
        if (!lenPrefix(2, len)) break;
        return array(pos + 3, len);
    case 0xdd:
        if (!lenPrefix(4, len)) break;
        return array(pos + 5, len);
    case 0xde:
        if (!lenPrefix(2, len)) break;
        return map(pos + 3, len);
    case 0xdf:
        if (!lenPrefix(4, len)) break;
        return map(pos + 5, len);
    default:
        return {CodecStatus::Error, IrspNull{}, pos}; // ext / 保留字节：不支持
    }
    return {CodecStatus::Incomplete, IrspNull{}, pos};
}

} // namespace

void MsgpackCodec::encode(const IrspValue &value, std::string &out) { encodeInto(value, out); }

std::string MsgpackCodec::encode(const IrspValue &value) {
    std::string out;
    encodeInto(value, out);
    return out;
}

CodecDecode MsgpackCodec::decode(std::string_view buffer) {
    PResult r = parseOne(buffer, 0);
    return CodecDecode{r.status, std::move(r.value), r.pos};
}

// ============================ 跨编码门面 ============================

std::string encodeValue(Encoding enc, const IrspValue &value) {
    return enc == Encoding::Msgpack ? MsgpackCodec::encode(value) : Irsp1Codec::encode(value);
}

CodecDecode decodeValue(Encoding enc, std::string_view buffer) {
    if (enc == Encoding::Msgpack) {
        return MsgpackCodec::decode(buffer);
    }
    Irsp1Codec::DecodeResult r = Irsp1Codec::decode(buffer);
    const CodecStatus st = r.status == Irsp1Codec::Status::Ok          ? CodecStatus::Ok
                           : r.status == Irsp1Codec::Status::Incomplete ? CodecStatus::Incomplete
                                                                        : CodecStatus::Error;
    return CodecDecode{st, std::move(r.value), r.consumed};
}

} // namespace irsp

File: sdk/irsp-client/Python/src/irsp_client/msgpack_codec.py
"""msgpack 编解码（IRSP V2 编码层的纯 Python 实现）。

对外接口与 ``irsp1.py`` 一致（``encode_request`` / ``decode``），故 client 层切换编码
几乎无差别。关键差异：

- **数值 value 由服务端编为 msgpack 原生 int/float**，客户端 *无需* 再按 ``type``
  解小端字节（对比 irsp1 必须 ``decode_value(type, raw)``）。
- 字符串解码为 ``str``（irsp1 的 bulk 解为 ``bytes``）。
- 错误为 ``{err, msg}`` map；本模块在 *顶层* 自动转成 ``IrspError``，与 irsp1 行为一致。

详见 irsp/doc/encoding/msgpack.md。一个 WS 二进制消息 = 一个 IRSP 帧（一个顶层值）。
"""
from __future__ import annotations

import struct
from typing import Any, List, Tuple, Union

from .irsp1 import IrspError  # 复用同一异常类型，client 层对两种编码无差别处理

# 解码后的值：str / int / float / bytes / None / IrspError / list / dict。
IrspValue = Union[str, int, float, bytes, None, IrspError, list, dict]


# ============================ 编码 ============================

def _enc_int(v: int, out: bytearray) -> None:
    if 0 <= v <= 0x7f:
        out.append(v)  # positive fixint
    elif -32 <= v < 0:
        out.append(v & 0xff)  # negative fixint
    elif 0 <= v <= 0xff:
        out += bytes((0xcc, v))
    elif 0 <= v <= 0xffff:
        out.append(0xcd); out += struct.pack(">H", v)
    elif 0 <= v <= 0xffffffff:
        out.append(0xce); out += struct.pack(">I", v)
    elif 0 <= v <= 0xffffffffffffffff:
        out.append(0xcf); out += struct.pack(">Q", v)
    elif -0x80 <= v:
        out.append(0xd0); out += struct.pack(">b", v)
    elif -0x8000 <= v:
        out.append(0xd1); out += struct.pack(">h", v)
    elif -0x80000000 <= v:
        out.append(0xd2); out += struct.pack(">i", v)
    else:
        out.append(0xd3); out += struct.pack(">q", v)


def _enc_str(b: bytes, out: bytearray) -> None:
    n = len(b)
    if n <= 31:
        out.append(0xa0 | n)
    elif n <= 0xff:
        out.append(0xd9); out.append(n)
    elif n <= 0xffff:
        out.append(0xda); out += struct.pack(">H", n)
    else:
        out.append(0xdb); out += struct.pack(">I", n)
    out += b


def _enc_bin(b: bytes, out: bytearray) -> None:
    n = len(b)
    if n <= 0xff:
        out.append(0xc4); out.append(n)
    elif n <= 0xffff:
        out.append(0xc5); out += struct.pack(">H", n)
    else:
        out.append(0xc6); out += struct.pack(">I", n)
    out += b


def _enc(obj: Any, out: bytearray) -> None:
    if obj is None:
        out.append(0xc0)
    elif obj is True:
        out.append(0xc3)
    elif obj is False:
        out.append(0xc2)
    elif isinstance(obj, int):
        _enc_int(obj, out)
    elif isinstance(obj, float):
        out.append(0xcb); out += struct.pack(">d", obj)
    elif isinstance(obj, str):
        _enc_str(obj.encode("utf-8"), out)
    elif isinstance(obj, (bytes, bytearray)):
        _enc_bin(bytes(obj), out)
    elif isinstance(obj, (list, tuple)):
        _enc_arr_hdr(len(obj), out)
        for it in obj:
            _enc(it, out)
    elif isinstance(obj, dict):
        _enc_map_hdr(len(obj), out)
        for k, v in obj.items():
            _enc(k, out)
            _enc(v, out)
    else:
        raise TypeError(f"msgpack: 不支持的类型 {type(obj).__name__}")


def _enc_arr_hdr(n: int, out: bytearray) -> None:
    if n <= 15:
        out.append(0x90 | n)
    elif n <= 0xffff:
        out.append(0xdc); out += struct.pack(">H", n)
    else:
        out.append(0xdd); out += struct.pack(">I", n)


def _enc_map_hdr(n: int, out: bytearray) -> None:
    if n <= 15:
        out.append(0x80 | n)
    elif n <= 0xffff:
        out.append(0xde); out += struct.pack(">H", n)
    else:
        out.append(0xdf); out += struct.pack(">I", n)


def encode(obj: Any) -> bytes:
    """把任意值编码为 msgpack 字节（含 map/array，便于离线构造样例帧）。"""
    out = bytearray()
    _enc(obj, out)
    return bytes(out)


def encode_request(parts: List[Union[str, bytes]]) -> bytes:
    """把命令编码为 msgpack 请求帧（数组）。字符串→str，原始字节→bin。"""
    out = bytearray()
    _enc_arr_hdr(len(parts), out)
    for p in parts:
        if isinstance(p, str):
            _enc_str(p.encode("utf-8"), out)
        elif isinstance(p, (bytes, bytearray)):
            _enc_bin(bytes(p), out)
        else:
            _enc_str(str(p).encode("utf-8"), out)
    return bytes(out)


# ============================ 解码 ============================

def _take_str(data: bytes, i: int, n: int) -> Tuple[str, int]:
    return data[i:i + n].decode("utf-8"), i + n


def _take_bin(data: bytes, i: int, n: int) -> Tuple[bytes, int]:
    return bytes(data[i:i + n]), i + n


def _parse_arr(data: bytes, i: int, n: int) -> Tuple[list, int]:
    out: list = []
    for _ in range(n):
        v, i = _parse(data, i)
        out.append(v)
    return out, i


def _parse_map(data: bytes, i: int, n: int) -> Tuple[dict, int]:
    obj: dict = {}
    for _ in range(n):
        k, i = _parse(data, i)
        v, i = _parse(data, i)
        if isinstance(k, (bytes, bytearray)):
            k = bytes(k).decode("utf-8")
        obj[k] = v
    return obj, i


def _parse(data: bytes, i: int) -> Tuple[Any, int]:
    c = data[i]
    i += 1
    if c <= 0x7f:
        return c, i  # positive fixint
    if c >= 0xe0:
        return c - 0x100, i  # negative fixint
    if 0x80 <= c <= 0x8f:
        return _parse_map(data, i, c & 0x0f)
    if 0x90 <= c <= 0x9f:
        return _parse_arr(data, i, c & 0x0f)
    if 0xa0 <= c <= 0xbf:
        return _take_str(data, i, c & 0x1f)
    if c == 0xc0:
        return None, i
    if c == 0xc2:
        return False, i
    if c == 0xc3:
        return True, i
    if c == 0xcc:
        return data[i], i + 1
    if c == 0xcd:
        return struct.unpack_from(">H", data, i)[0], i + 2
    if c == 0xce:
        return struct.unpack_from(">I", data, i)[0], i + 4
    if c == 0xcf:
        return struct.unpack_from(">Q", data, i)[0], i + 8
    if c == 0xd0:
        return struct.unpack_from(">b", data, i)[0], i + 1
    if c == 0xd1:
        return struct.unpack_from(">h", data, i)[0], i + 2
    if c == 0xd2:
        return struct.unpack_from(">i", data, i)[0], i + 4
    if c == 0xd3:
        return struct.unpack_from(">q", data, i)[0], i + 8
    if c == 0xca:
        return struct.unpack_from(">f", data, i)[0], i + 4
    if c == 0xcb:
        return struct.unpack_from(">d", data, i)[0], i + 8
    if c == 0xc4:
        return _take_bin(data, i + 1, data[i])
    if c == 0xc5:
        return _take_bin(data, i + 2, struct.unpack_from(">H", data, i)[0])
    if c == 0xc6:
        return _take_bin(data, i + 4, struct.unpack_from(">I", data, i)[0])
    if c == 0xd9:
        return _take_str(data, i + 1, data[i])
    if c == 0xda:
        return _take_str(data, i + 2, struct.unpack_from(">H", data, i)[0])
    if c == 0xdb:
        return _take_str(data, i + 4, struct.unpack_from(">I", data, i)[0])
    if c == 0xdc:
        return _parse_arr(data, i + 2, struct.unpack_from(">H", data, i)[0])
    if c == 0xdd:
        return _parse_arr(data, i + 4, struct.unpack_from(">I", data, i)[0])
    if c == 0xde:
        return _parse_map(data, i + 2, struct.unpack_from(">H", data, i)[0])
    if c == 0xdf:
        return _parse_map(data, i + 4, struct.unpack_from(">I", data, i)[0])
    raise ValueError(f"msgpack: 未知首字节 0x{c:02x}")


def decode(data: bytes) -> IrspValue:
    """解码一个完整 msgpack 帧。顶层 ``{err,msg}`` map 自动转为 IrspError。"""
    value, _ = _parse(data, 0)
    if isinstance(value, dict) and "err" in value:
        code = value.get("err")
        msg = value.get("msg", "")
        return IrspError(str(code), str(msg))
    return value
